### PR TITLE
Adding an option to restrict the access in course catalog API

### PR DESCRIPTION
**Pull request by manikarthikk**
#### What does this PR do? Please provide some context
This will introduce new setting "RESTRICT_COURSES_API" in courses API which needs to be added in site configurations in admin panel to restrict  the api for a specific site. 
#### Where should the reviewer start?
deploy the code and add RESTRICT_COURSES_API to true in admin panel site_configurations and we can't access the site and throws error to users . url: "api/courses/v1/courses"
#### How can this be manually tested? (brief repro steps)
add this RESTRICT_COURSES_API  in admin panel and we can see the api url is restricted. and can be accessed with  access token with staff access 
#### Definition of done:
- [x] Title of the pull request is clear and informative
- [ ] Add pull request hyperlink to relevant TFS items
- [x] For large or complex change: schedule an in-person review session
- [x] This change has appropriate test coverage

#### Reminders BEFORE merging
1. Get at least two approvals
1. If you're merging from a feature branch into the development branch then "flatten" or "squash" commits
1. If merging from the development branch into master (or porting changes from upstream) then use github's UI to get review feedback, but use the git command line interface to complete the actual merge.

#### Reminders AFTER merging
1. Delete the remote feature branch
1. Resolve relevant TFS items
1. (reverse merge) If you merged from the development branch into master then check to see if there are any changes in master that can be merged down to the development branch (like hotfixes, etc). In this case, use github's UI for feedback and the git command line interface for the actual merge.

[//]: # ( todo: If you merged into development branch then verify change in our "rolling deployment" environment. Then notify stakeholders interested in or involved with the change )

[//]: # ( fyi: This content was heavily inspired by )
[//]: # ( 1 Our team's policies and processes )
[//]: # ( 2 https://github.com/sprintly/sprint.ly-culture/blob/master/pr-template.md )
[//]: # ( 3 The book "The Checklist Manifesto: How to Get Things Right" by Atul Gawande )
[//]: # ( 4 https://github.com/Azure/azure-event-hubs/blob/master/.github/PULL_REQUEST_TEMPLATE.md )

### CODE REVIEW

File: lms/djangoapps/course_api/tests/test_views.py [hunk 1]
```diff
@@ -10,6 +10,11 @@
 from xmodule.modulestore.tests.django_utils import SharedModuleStoreTestCase, ModuleStoreTestCase
 from .mixins import CourseApiFactoryMixin, TEST_PASSWORD
 from ..views import CourseDetailView
+from openedx.core.djangoapps.site_configuration.tests.test_util import with_site_configuration
+
+TEST_SITE_CONFIGURATION = {
```
Comment (btelnes) [inline lms/djangoapps/course_api/tests/test_views.py]: Perhaps externalize the configuration?

File: lms/djangoapps/course_api/views.py [hunk 2]
```diff
@@ -196,9 +197,12 @@ def get_queryset(self):
         """
         Return a list of courses visible to the user.
         """
+        restrict_course_api = configuration_helpers.get_value('RESTRICT_COURSES_API')
         form = CourseListGetForm(self.request.query_params, initial={'requesting_user': self.request.user})
         if not form.is_valid():
             raise ValidationError(form.errors)
+        elif restrict_course_api and not self.request.user.is_staff:
+            raise PermissionDenied()
```
Comment (kabirkhan) [inline lms/djangoapps/course_api/views.py]: I don't think this should be limited to just global staff users. 

A user can be a course staff member, global staff or global superuser and those roles should probably have access here.

Check out the lms/djangoapps/grades/api/views.py _get_effective_user function for validating course staff / global staff access.

Comment (manikarthikk) [reply lms/djangoapps/course_api/views.py]: Had discussion with @kabirkhan on the same and he is fine with current implementation.

File: lms/djangoapps/course_api/tests/test_views.py (hunk 1 above)
Comment (manikarthikk) [reply lms/djangoapps/course_api/tests/test_views.py]: just followed the existing pattern in other tests

File: lms/djangoapps/course_api/views.py (hunk 2 above)
Comment (manikarthikk) [reply lms/djangoapps/course_api/views.py]: 1) we don't need to  give special permissions to superuser explicitly for accessing the functions as per implementation.
2) the _get_effective_user function  is available in ginkgo and not available in ficus moreover, here we are able to know the user's role by request.user method.

**Pull request merged**